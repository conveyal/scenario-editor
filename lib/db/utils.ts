import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollectionResults(obj: unknown) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollectionResults)
  if (typeof obj === 'object') return mapValues(obj, serializeCollectionResults)
  return obj
}
