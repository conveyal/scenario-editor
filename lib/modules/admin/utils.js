import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect(Component) {
  return connect(createStructuredSelector(select), actions)(Component)
}

const add0 = n => (n >= 10 ? n : `0${n}`)

export function msToDuration(ms) {
  const durationS = Math.floor((Date.now() - ms) / 1000)
  const s = durationS % 60
  const m = Math.floor((durationS / 60) % 60)
  const h = Math.floor((durationS / 60 / 60) % 60)

  return `${add0(h)}:${add0(m)}:${add0(s)}`
}

/**
 * Example commit string: "v3.4.0-118-gbe7199c". Take the last section after
 * splitting on the hyphens and remove the "g".
 */
const githubUrl = 'https://github.com/conveyal/r5/commits'
export function createR5Url(workerCommit) {
  // not a release version
  if (workerCommit.indexOf('-') !== -1) {
    const hash = workerCommit.split('-')[2].slice(1) //
    return `${githubUrl}/${hash}`
  }
  return `${githubUrl}/${workerCommit}`
}

export function createWorkerUrl(instanceId, region) {
  return `https://${region}.console.aws.amazon.com/ec2/v2/home?region=${region}#Instances:instanceId=${instanceId};sort=tag:Name`
}

const C1 = '#2389c9'
const C2 = '#36c923'
const C3 = '#aab4bb'

/**
 * Create a progress bar with a linear-gradient.
 */
export function createLinearGradientForJob(c, d, t) {
  const p1 = (c / t) * 100
  const p2 = (d / t) * 100
  return `linear-gradient(to right,
    ${C1} 0%,
    ${C1} ${p1}%,
    ${C2} ${p1}%,
    ${C2} ${p2}%,
    ${C3} ${p2}%,
    ${C3} 100%)`
}
