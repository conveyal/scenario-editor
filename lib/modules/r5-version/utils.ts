import {
  MINIMUM_R5_VERSION,
  VERSION_PARSE_REGEX,
  OLD_VERSION_PARSE_REGEX
} from './constants'

/**
 * Convert an r5 version to a number for comparing
 * NB Assumes no major, minor, patch, or commit version exceeds 100
 */
export function versionToNumber(version?: string): number {
  const versionString = version || ''
  if (versionString.match(/\./g) && versionString.match(/\./g).length == 1) {
    const versionMatch = versionString.match(VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [, major = '0', minor = '0', commit = '0'] = versionMatch

      return (
        parseInt(major) * 10000 + parseInt(minor) * 100 + parseInt(commit) / 100
      )
    }

    return 0
  } else {
    const versionMatch = versionString.match(OLD_VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [
        ,
        major = '0',
        minor = '0',
        patch = '0',
        commit = '0'
      ] = versionMatch

      return (
        parseInt(major) * 10000 +
        parseInt(minor) * 100 +
        parseInt(patch) +
        parseInt(commit) / 100
      )
    }

    return 0
  }
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = (version || '').search(/-/g) !== -1
  return isTestVersion || workerVersionInRange(version, min, max)
}
