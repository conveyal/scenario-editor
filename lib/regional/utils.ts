function getDefaultCutoff(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.cutoffsMinutes))
    return a.cutoffsMinutes[Math.floor(a.cutoffsMinutes.length / 2)]
  return a.cutoffMinutes
}

function getDefaultDestinationPointSet(a: CL.RegionalAnalysis): string | void {
  const ids = a.destinationPointSetIds
  if (Array.isArray(ids)) return ids[0]
}

function getDefaultPercentile(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.travelTimePercentiles))
    return a.travelTimePercentiles[
      Math.floor(a.travelTimePercentiles.length / 2)
    ]
  return a.travelTimePercentile
}

/**
 * Choose
 */
export function getDefaultVariants(analysis: CL.RegionalAnalysis) {
  return {
    cutoff: getDefaultCutoff(analysis),
    destinationPointSetId: getDefaultDestinationPointSet(analysis),
    percentile: getDefaultPercentile(analysis)
  }
}

export function getComparisonVariants(comparison: CL.RegionalAnalysis) {
  return {
    comparisonCutoff: getDefaultCutoff(comparison),
    comparisonDestinationPointSetId: getDefaultDestinationPointSet(comparison),
    comparisonPercentile: getDefaultPercentile(comparison)
  }
}

/**
 * Don't compare incomparable analyses.
 */
export function getValidComparisonAnalyses(
  allAnalyses: CL.RegionalAnalysis[],
  activeJobs: CL.RegionalJob[],
  analysis: CL.RegionalAnalysis
) {
  return allAnalyses.filter(
    (c) =>
      activeJobs.findIndex((job) => job.jobId === c._id) === -1 &&
      c.request.originPointSetKey == null &&
      c.zoom === analysis.zoom &&
      c.width === analysis.width &&
      c.height === analysis.height &&
      c.north === analysis.north &&
      c.west === analysis.west
  )
}
