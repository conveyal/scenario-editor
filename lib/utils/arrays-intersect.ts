// Do the arrays intersect?
export default function arraysIntersect<T>(a1: T[], a2: T[]) {
  for (const v of a1) {
    if (a2.indexOf(v) > -1) return true
  }
  return false
}
