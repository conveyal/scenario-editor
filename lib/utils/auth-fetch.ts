import {parse} from 'cookie'

import {safeFetch, SafeResponse} from './safe-fetch'

export function createAuthHeaders(user?: CL.User) {
  const headers = {
    Authorization: `bearer ${user?.idToken}`
  }

  // Add the admin access group for administrators if it exists.
  if (user?.accessGroup === process.env.NEXT_PUBLIC_ADMIN_ACCESS_GROUP) {
    const adminTempAccessGroup = parse(document.cookie).adminTempAccessGroup
    if (adminTempAccessGroup?.length > 0) {
      headers['X-Conveyal-Access-Group'] = adminTempAccessGroup
    }
  }
  return headers
}

/**
 * Fetch wrapper that includes authentication. Defaults headers to JSON.
 */
export default function authFetch<T>(
  url: string,
  user: CL.User,
  parser: (res: Response) => Promise<T> = (res) => res.json(),
  options?: RequestInit
): Promise<SafeResponse<T>> {
  return safeFetch(url, parser, {
    mode: 'cors',
    ...options,
    headers: {
      ...createAuthHeaders(user),
      ...options?.headers
    }
  })
}
