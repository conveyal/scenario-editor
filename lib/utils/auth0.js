// @flow
import message from '@conveyal/woonerf/message'
import Auth0Lock from 'auth0-lock'

const CLIENT_ID = process.env.AUTH0_CLIENT_ID
const DOMAIN = process.env.AUTH0_DOMAIN

export const authIsRequired = CLIENT_ID && DOMAIN

const localStorage = window && window.localStorage
  ? window.localStorage
  : {
    getItem () {},
    setItem () {}
  }

const lock = authIsRequired
  ? new Auth0Lock(CLIENT_ID, DOMAIN, {
    auth: {
      params: {
        scope: 'openid email profile analyst offline_access'
      },
      redirect: false
    },
    allowSignUp: false,
    theme: {
      logo: 'https://s3-eu-west-1.amazonaws.com/analyst-logos/conveyal-128x128.png',
      primaryColor: '#2389c9'
    },
    closable: false,
    languageDictionary: {
      title: message('authentication.logIn')
    }
  })
  : {
    checkSession () {},
    getUserInfo () {},
    hide () {},
    logout () {},
    on () {},
    show () {}
  }

function flashError (error) {
  console.error(error)
  lock.show({
    flashMessage: {
      type: 'error',
      text: error.error_description
    }
  })
}

// Set the default handler for auth errors
lock.on('authorization_error', flashError)
lock.on('unrecoverable_error', flashError)

export function login (next: (any) => void) {
  function handleAuthResult (authResult) {
    console.log('handleAuthResult', authResult)
    lock.getUserInfo(authResult.accessToken, (error, profile) => {
      if (error) {
        localStorage.removeItem('user')
        lock.show({
          flashMessage: {
            type: 'error',
            text: error.error_description
          }
        })
      } else {
        localStorage.setItem('user', JSON.stringify(profile))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      }
    })
  }

  lock.checkSession({}, (error, authResult) => {
    console.log('checkSession', error, authResult)
    if (error || !authResult) {
      lock.show()
      lock.on('authenticated', handleAuthResult)
    } else {
      try {
        const profile = JSON.parse(localStorage.getItem('user'))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      } catch (e) {
        console.error('error processing user profile from localStorage', e)
        handleAuthResult(authResult)
      }
    }
  })
}

export function logout () {
  localStorage.removeItem('user')
  lock.logout({returnTo: `${window.location.origin}/login`})
}
