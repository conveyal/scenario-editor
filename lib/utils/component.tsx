import {UserProvider} from '@auth0/nextjs-auth0'
import {ThemeProvider, extendTheme} from '@chakra-ui/react'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {localUser} from 'lib/user'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

const theme = extendTheme({})

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <UserProvider user={localUser}>
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <RouterContext.Provider value={router}>
            <Map>{props.children}</Map>
          </RouterContext.Provider>
        </ThemeProvider>
      </Provider>
    </UserProvider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Helper function to mount and unmount a component.
 */
export function testComponentMount(Component, props) {
  test(`Component mount(${getName(Component)})`, () => {
    const c = testComponent(Component, props)
    const t = c.mount()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
