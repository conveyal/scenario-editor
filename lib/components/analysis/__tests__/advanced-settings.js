import {testComponent} from 'lib/utils/component'

import {
  mockProfileRequest,
  mockRegion,
  mockRegionalAnalyses
} from 'lib/utils/mock-data'

import AdvancedSettings from '../advanced-settings'

describe('Components > Analysis > Advanced Settings', () => {
  const props = {
    disabled: false,
    profileRequest: mockProfileRequest,
    regionalAnalyses: mockRegionalAnalyses,
    regionBounds: mockRegion.bounds
  }
  it('should render correctly', () => {
    const tree = testComponent(AdvancedSettings, {...props})
    const mounted = tree.mount()
    mounted.unmount()
  })
})
