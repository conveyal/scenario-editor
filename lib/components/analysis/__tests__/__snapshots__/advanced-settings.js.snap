// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components > Analysis > Advanced Settings should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AdvancedSettings
    disabled={false}
    hideBoundsEditor={[MockFunction]}
    profileRequest={
      Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bounds": Object {
          "east": -76,
          "north": 39,
          "south": 38,
          "west": -77,
        },
        "bundleId": "1",
        "date": "2016-01-16",
        "defaultExpanded": true,
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromLat": 12,
        "fromLon": 34,
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxRides": 4,
        "maxTripDurationMinutes": 60,
        "maxWalkTime": 20,
        "monteCarloDraws": 200,
        "name": "Profile Request Test",
        "projectId": "1",
        "reachabilityThreshold": 0,
        "regionId": "1",
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "travelTimePercentile": 50,
        "variantIndex": 0,
        "walkSpeed": 1.3888888888888888,
        "workerVersion": "v4.0.0",
      }
    }
    regionBounds={
      Object {
        "east": -76,
        "north": 39,
        "south": 38,
        "west": -77,
      }
    }
    regionalAnalyses={Array []}
    setProfileRequest={[MockFunction]}
    showBoundsEditor={[MockFunction]}
  >
    <section
      className="panel panel-default inner-panel advanced-panel"
    >
      <a
        className="panel-heading clearfix"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        tabIndex={0}
      >
        <Icon
          type="gear"
        >
          <i
            className="fa fa-gear fa-fw "
          />
        </Icon>
        <strong>
          Advanced Settings
        </strong>
        <Icon
          className="pull-right"
          type="caret-right"
        >
          <i
            className="fa fa-caret-right fa-fw pull-right"
          />
        </Icon>
      </a>
    </section>
  </AdvancedSettings>
</Provider>
`;

exports[`Components > Analysis > Advanced Settings should render correctly with expanded advanced settings 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AdvancedSettings
    disabled={false}
    hideBoundsEditor={[MockFunction]}
    profileRequest={
      Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bounds": Object {
          "east": -76,
          "north": 39,
          "south": 38,
          "west": -77,
        },
        "bundleId": "1",
        "date": "2016-01-16",
        "defaultExpanded": true,
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromLat": 12,
        "fromLon": 34,
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxRides": 4,
        "maxTripDurationMinutes": 60,
        "maxWalkTime": 20,
        "monteCarloDraws": 200,
        "name": "Profile Request Test",
        "projectId": "1",
        "reachabilityThreshold": 0,
        "regionId": "1",
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "travelTimePercentile": 50,
        "variantIndex": 0,
        "walkSpeed": 1.3888888888888888,
        "workerVersion": "v4.0.0",
      }
    }
    regionBounds={
      Object {
        "east": -76,
        "north": 39,
        "south": 38,
        "west": -77,
      }
    }
    regionalAnalyses={Array []}
    setProfileRequest={[MockFunction]}
    showBoundsEditor={[MockFunction]}
  >
    <section
      className="panel panel-default inner-panel advanced-panel"
    >
      <a
        className="panel-heading clearfix"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        tabIndex={0}
      >
        <Icon
          type="gear"
        >
          <i
            className="fa fa-gear fa-fw "
          />
        </Icon>
        <strong>
          Advanced Settings
        </strong>
        <Icon
          className="pull-right"
          type="caret-down"
        >
          <i
            className="fa fa-caret-down fa-fw pull-right"
          />
        </Icon>
      </a>
      <div
        className="panel-body"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-4"
          >
            <EgressModeSelector
              disabled={false}
              egressModes="WALK"
              update={[MockFunction]}
            >
              <Group
                label="Egress Modes"
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="egress-modes-input-0"
                  >
                    Egress Modes
                  </label>
                  <br />
                  <Group
                    disabled={false}
                  >
                    <div
                      className="btn-group"
                      disabled={false}
                    >
                      <Button
                        active={true}
                        onClick={[Function]}
                        title="Walk"
                      >
                        <a
                          className="btn btn-default active"
                          href="#"
                          onClick={[Function]}
                          tabIndex={0}
                          title="Walk"
                        >
                          <Icon
                            type="male"
                          >
                            <i
                              className="fa fa-male fa-fw "
                            />
                          </Icon>
                        </a>
                      </Button>
                      <Button
                        active={false}
                        disabled={true}
                        onClick={[Function]}
                        title="CURRENTLY DISABLED: Bike"
                      >
                        <a
                          className="btn btn-default"
                          disabled={true}
                          href="#"
                          onClick={[Function]}
                          tabIndex={0}
                          title="CURRENTLY DISABLED: Bike"
                        >
                          <Icon
                            type="bicycle"
                          >
                            <i
                              className="fa fa-bicycle fa-fw "
                            />
                          </Icon>
                        </a>
                      </Button>
                    </div>
                  </Group>
                </div>
              </Group>
            </EgressModeSelector>
          </div>
          <div
            className="col-xs-4"
          >
            <Number
              disabled={false}
              label="Walk Speed (km/h)"
              max={15}
              min={3}
              name="walkSpeed"
              onChange={[Function]}
              step={0.1}
              value={5}
            >
              <Group
                disabled={false}
                id="walk-speed-input-1"
                label="Walk Speed (km/h)"
                max={15}
                min={3}
                name="walkSpeed"
                onChange={[Function]}
                step={0.1}
                value={5}
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="walk-speed-input-1"
                  >
                    Walk Speed (km/h)
                  </label>
                  <Input
                    disabled={false}
                    id="walk-speed-input-1"
                    label="Walk Speed (km/h)"
                    max={15}
                    min={3}
                    name="walkSpeed"
                    onChange={[Function]}
                    onWheel={[Function]}
                    step={0.1}
                    type="number"
                    value={5}
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      id="walk-speed-input-1"
                      max={15}
                      min={3}
                      name="walkSpeed"
                      onChange={[Function]}
                      onWheel={[Function]}
                      placeholder="Walk Speed (km/h)"
                      step={0.1}
                      type="number"
                      value={5}
                    />
                  </Input>
                </div>
              </Group>
            </Number>
          </div>
          <div
            className="col-xs-4"
          >
            <Number
              disabled={false}
              label="Bike Speed (km/h)"
              max={20}
              min={5}
              name="bikeSpeed"
              onChange={[Function]}
              step={0.1}
              value={15}
            >
              <Group
                disabled={false}
                id="bike-speed-input-2"
                label="Bike Speed (km/h)"
                max={20}
                min={5}
                name="bikeSpeed"
                onChange={[Function]}
                step={0.1}
                value={15}
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="bike-speed-input-2"
                  >
                    Bike Speed (km/h)
                  </label>
                  <Input
                    disabled={false}
                    id="bike-speed-input-2"
                    label="Bike Speed (km/h)"
                    max={20}
                    min={5}
                    name="bikeSpeed"
                    onChange={[Function]}
                    onWheel={[Function]}
                    step={0.1}
                    type="number"
                    value={15}
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      id="bike-speed-input-2"
                      max={20}
                      min={5}
                      name="bikeSpeed"
                      onChange={[Function]}
                      onWheel={[Function]}
                      placeholder="Bike Speed (km/h)"
                      step={0.1}
                      type="number"
                      value={15}
                    />
                  </Input>
                </div>
              </Group>
            </Number>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-xs-4"
          />
          <div
            className="col-xs-4"
          >
            <Number
              disabled={false}
              label="Max Walk Time (min)"
              min={1}
              name="maxWalkTime"
              onChange={[Function]}
              step={1}
              value={20}
            >
              <Group
                disabled={false}
                id="max-walk-time-input-3"
                label="Max Walk Time (min)"
                min={1}
                name="maxWalkTime"
                onChange={[Function]}
                step={1}
                value={20}
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="max-walk-time-input-3"
                  >
                    Max Walk Time (min)
                  </label>
                  <Input
                    disabled={false}
                    id="max-walk-time-input-3"
                    label="Max Walk Time (min)"
                    min={1}
                    name="maxWalkTime"
                    onChange={[Function]}
                    onWheel={[Function]}
                    step={1}
                    type="number"
                    value={20}
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      id="max-walk-time-input-3"
                      min={1}
                      name="maxWalkTime"
                      onChange={[Function]}
                      onWheel={[Function]}
                      placeholder="Max Walk Time (min)"
                      step={1}
                      type="number"
                      value={20}
                    />
                  </Input>
                </div>
              </Group>
            </Number>
          </div>
          <div
            className="col-xs-4"
          >
            <Number
              disabled={false}
              label="Max Bike Time (min)"
              min={1}
              name="maxBikeTime"
              onChange={[Function]}
              step={1}
              value={20}
            >
              <Group
                disabled={false}
                id="max-bike-time-input-4"
                label="Max Bike Time (min)"
                min={1}
                name="maxBikeTime"
                onChange={[Function]}
                step={1}
                value={20}
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="max-bike-time-input-4"
                  >
                    Max Bike Time (min)
                  </label>
                  <Input
                    disabled={false}
                    id="max-bike-time-input-4"
                    label="Max Bike Time (min)"
                    min={1}
                    name="maxBikeTime"
                    onChange={[Function]}
                    onWheel={[Function]}
                    step={1}
                    type="number"
                    value={20}
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      id="max-bike-time-input-4"
                      min={1}
                      name="maxBikeTime"
                      onChange={[Function]}
                      onWheel={[Function]}
                      placeholder="Max Bike Time (min)"
                      step={1}
                      type="number"
                      value={20}
                    />
                  </Input>
                </div>
              </Group>
            </Number>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-xs-6"
          >
            <Number
              disabled={false}
              label="Maximum transfers"
              max={7}
              min={0}
              name="maxTransfers"
              onChange={[Function]}
              step={1}
              value={3}
            >
              <Group
                disabled={false}
                id="max-transfers-input-5"
                label="Maximum transfers"
                max={7}
                min={0}
                name="maxTransfers"
                onChange={[Function]}
                step={1}
                value={3}
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="max-transfers-input-5"
                  >
                    Maximum transfers
                  </label>
                  <Input
                    disabled={false}
                    id="max-transfers-input-5"
                    label="Maximum transfers"
                    max={7}
                    min={0}
                    name="maxTransfers"
                    onChange={[Function]}
                    onWheel={[Function]}
                    step={1}
                    type="number"
                    value={3}
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      id="max-transfers-input-5"
                      max={7}
                      min={0}
                      name="maxTransfers"
                      onChange={[Function]}
                      onWheel={[Function]}
                      placeholder="Maximum transfers"
                      step={1}
                      type="number"
                      value={3}
                    />
                  </Input>
                </div>
              </Group>
            </Number>
          </div>
          <div
            className="col-xs-6"
          >
            <Number
              disabled={false}
              label="Simulated schedules"
              max={10000}
              min={1}
              name="monteCarloDraws"
              onChange={[Function]}
              step={1}
              value={200}
            >
              <Group
                disabled={false}
                id="monte-carlo-draws-input-6"
                label="Simulated schedules"
                max={10000}
                min={1}
                name="monteCarloDraws"
                onChange={[Function]}
                step={1}
                value={200}
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="monte-carlo-draws-input-6"
                  >
                    Simulated schedules
                  </label>
                  <Input
                    disabled={false}
                    id="monte-carlo-draws-input-6"
                    label="Simulated schedules"
                    max={10000}
                    min={1}
                    name="monteCarloDraws"
                    onChange={[Function]}
                    onWheel={[Function]}
                    step={1}
                    type="number"
                    value={200}
                  >
                    <input
                      className="form-control"
                      disabled={false}
                      id="monte-carlo-draws-input-6"
                      max={10000}
                      min={1}
                      name="monteCarloDraws"
                      onChange={[Function]}
                      onWheel={[Function]}
                      placeholder="Simulated schedules"
                      step={1}
                      type="number"
                      value={200}
                    />
                  </Input>
                </div>
              </Group>
            </Number>
          </div>
        </div>
        <Group
          label="Bounds of analysis"
        >
          <div
            className="form-group"
          >
            <label
              htmlFor="bounds-of-analysis-input-7"
            >
              Bounds of analysis
            </label>
            <div
              className="row"
            >
              <div
                className="col-xs-6"
              >
                <Select
                  arrowRenderer={[Function]}
                  autosize={true}
                  backspaceRemoves={true}
                  backspaceToRemoveMessage="Press backspace to remove {label}"
                  clearAllText="Clear all"
                  clearRenderer={[Function]}
                  clearValueText="Clear value"
                  clearable={false}
                  closeOnSelect={true}
                  deleteRemoves={true}
                  delimiter=","
                  disabled={false}
                  escapeClearsValue={true}
                  filterOptions={[Function]}
                  ignoreAccents={true}
                  ignoreCase={true}
                  inputProps={Object {}}
                  isLoading={false}
                  joinValues={false}
                  labelKey="label"
                  matchPos="any"
                  matchProp="any"
                  menuBuffer={0}
                  menuRenderer={[Function]}
                  multi={false}
                  noResultsText="No results found"
                  onBlurResetsInput={true}
                  onChange={[Function]}
                  onCloseResetsInput={true}
                  onSelectResetsInput={true}
                  openOnClick={true}
                  optionComponent={[Function]}
                  options={
                    Array [
                      Object {
                        "label": "Entire region",
                        "value": "__REGION",
                      },
                    ]
                  }
                  pageSize={5}
                  placeholder="Select..."
                  removeSelected={true}
                  required={false}
                  rtl={false}
                  scrollMenuIntoView={true}
                  searchable={true}
                  simpleValue={false}
                  tabSelectsValue={true}
                  trimFilter={true}
                  value="__REGION"
                  valueComponent={[Function]}
                  valueKey="value"
                >
                  <div
                    className="Select has-value is-searchable Select--single"
                  >
                    <div
                      className="Select-control"
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                    >
                      <span
                        className="Select-multi-value-wrapper"
                        id="react-select-2--value"
                      >
                        <Value
                          disabled={false}
                          id="react-select-2--value-item"
                          instancePrefix="react-select-2-"
                          onClick={null}
                          placeholder="Select..."
                          value={
                            Object {
                              "label": "Entire region",
                              "value": "__REGION",
                            }
                          }
                        >
                          <div
                            className="Select-value"
                          >
                            <span
                              aria-selected="true"
                              className="Select-value-label"
                              id="react-select-2--value-item"
                              role="option"
                            >
                              Entire region
                            </span>
                          </div>
                        </Value>
                        <AutosizeInput
                          aria-activedescendant="react-select-2--value"
                          aria-expanded="false"
                          aria-haspopup="false"
                          aria-owns=""
                          className="Select-input"
                          injectStyles={true}
                          minWidth="5"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          required={false}
                          role="combobox"
                          value=""
                        />
                      </span>
                      <span
                        className="Select-arrow-zone"
                        onMouseDown={[Function]}
                      >
                        <span
                          className="Select-arrow"
                          onMouseDown={[Function]}
                        />
                      </span>
                    </div>
                  </div>
                </Select>
              </div>
              <div
                className="col-xs-6"
              >
                <Button
                  block={true}
                  disabled={false}
                  onClick={[Function]}
                  style="warning"
                >
                  <a
                    className="btn btn-warning btn-block"
                    disabled={false}
                    href="#"
                    onClick={[Function]}
                    tabIndex={0}
                  >
                    <Icon
                      type="pencil"
                    >
                      <i
                        className="fa fa-pencil fa-fw "
                      />
                    </Icon>
                     Set custom geographic bounds
                  </a>
                </Button>
              </div>
            </div>
          </div>
        </Group>
      </div>
    </section>
  </AdvancedSettings>
</Provider>
`;
